### ISSUE TITLE
FIXME - java

### ISSUE THREAD

**Issue by bitPogo** [akimd/bison#83]
Hello!
I am currently working on a small project, which utilises Bison for [Kotlin (KMP/KMM)](https://kotlinlang.org/docs/multiplatform.html).  
This far I managed to create a Gradle task and checked the Java Source for any obstacle, which makes the Java2Kotlin Converter do funny things and started to eliminate them. 
However while doing so I stumbled over this [FIXME](https://github.com/akimd/bison/blob/6571c2d1b134bd04d81f1426c08985407e8c7087/data/skeletons/lalr1.java#L997), which unfortunately does not provide any further information or pointers to any test case this breaks. Therefore it would be lovely if you point me to additional informations or give me a hint what is wrong there, so I am maybe able to fix it. Of course if I am able to fix it, I gladly provide any contribution to Bison.
However in the long run direct Kotlin support would be great, but I my knowledge around M4 is for now too limited be be of any help to make something  similar like [this project](https://github.com/iliabylich/rust-bison-skeleton).

**Comment by bokrenok** [comment]
Hello!
I've just run into this while investigating the way syntax errors are printed.
And I would like to share what I've found. 
(BTW my project is for C++, but the whole idea is the same and source code is very similar).

TL;DR;
When the mentioned line is reached by the code execution, it means that there were found some expected tokens present, but the allowed limit has been reached and the execution of the function `getExpectedTokens` would result in 'no possibilities found'... which could be quite an unexpected.

Full story:
So, I was investigating why in some cases syntax errors displays expecting alternatives like this:

`syntax error, unexpected =, expecting byte or le16 or be16 `

and in other cases there are no expecting alternatives displayed like this:

`syntax error, unexpected =`

It turned out that the function `getExpectedTokens` (or `expected_tokens` in C++) is intended to return no more than given `yyargn` possible alternatives. 
Whenever there are more possible alternatives - the function returns no alternatives at all.
The line you are referred to is exactly about it: if there are more than `yyargn` alternatives - none of them are returned at all.

And the `yyargn` value is set by `TOKENMAX = 5` constant value, which is defined right before `getExpectedTokens` call.
I suppose that for jave it is kind of legacy from original C/C++, which can be rewritten in a better way.

But anyway there is kind of a dilemma: what should be done in case of lots of possible alternatives?
Given that the list of them is to be printed in syntax error message for user.

1. should it return a huge list of alternatives (ending up whith quite a long enumeration of values ' ... or ... or ... or ... ')
2. should it return at least `yyargn` number of them (and loose info about other possibilities, basically restrict the list only to fixed amout of data)
3. should it return nothing at all (leaving the user without any key to correct values which are usable here)
and this is what it does right now and ever.

As I sought through the repo, I found that for the first time the constant `YYERROR_VERBOSE_ARGS_MAXIMUM` was introduced in this commit: https://github.com/akimd/bison/commit/a0ffc1751e712e55b27aa7349ec7db302557476b
and then rewritten with additional comments here  https://github.com/akimd/bison/commit/d2060f0634f4adfb5db74cce540a9d27806091fe
And earlier it was just `if (count < 5)`

**Comment by bitPogo** [comment]
Thx for the explanation and the pointers!

Yeah well...the dilemma is tricky in terms of usability.  1) is maybe ugly but gives you the most possible information and is open to refinement I suppose (like the famous `did you mean XXX` or at least nice formatting), while 3) is in the end no help at all in my eyes. That said, I think for Bison itself there should be a discussion before committing any thing, how to deal with it.

In general the Java Source could use quite some polishing even while Bison is maybe not so popular in realm of Java for some reasons. Also that could benefit its flexibility in terms of modification (which I really adore after spending some time with it aside from M4). And I bet that needs discussion as well upfront.

For my current use case I figured out (unfortunately), Bison is not right hammer for that nail. But what is maybe good to know - with light modification in the current Java Skeleton and some post processing once the source is converted to Kotlin it is possible to use it for KMP. (Of course only with a certain degree of certainty, since I did not made exhausted tests)

All that said I close the issue for now.
